test/main.js

```diff
@@ -173,32 +175,103 @@ const runner = async () => {
         if (driver) driver.finish()
       }
     })
-    console.log('Finished test')
   })
 
-  console.log('finished')
+  console.log(chalk.green('Tests finished'))
   process.exit(totalFailures > 0 ? 1 : 0)
 }
 
-const server = exec('npm run travis')
-const killServer = () => {
-  console.log('Killing server')
-  if (!server.killed) {
-    server.kill()
-    console.log('Kill signal sent')
-  } else {
-    console.log('Kill signal already sent')
+const killApp = (appProcess) => {
+  console.log(chalk.grey('Killing test app'))
+
+  if (!appProcess.killed) {
+    appProcess.kill()
+    console.log(chalk.green('Kill signal sent'))
   }
 }
 
-server.stdout.on('data', (data) => {
-  if (data.includes('Available on:')) {
-    runner()
+const killMockServer = (dockerContainerId) => {
+  console.log(chalk.grey('Killing mock server'))
+  exec(`docker stop ${dockerContainerId} -t0`, (error) => {
+    if (error) {
+      console.log(chalk.yellow('Error killing mock server:'), error)
+    } else {
+      console.log(chalk.green('Mock server killed'))
+    }
+  })
+}
+
+console.log(chalk.bold.green('Starting mock server'))
+
+const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms))
+
+const pingMockServer = () =>
+  new Promise((resolve, reject) => {
+    http
+      .get('http://localhost:8081/ping', (res) => {
+        res.on('data', () => {})
+        res.on('end', () => {})
+
+        if (res.statusCode === 200) {
+          resolve()
+        }
+      })
+      .on('error', reject)
+  })
+
+const waitForMockServer = async () => {
+  let isMockServerRunning = false
+
+  while (!isMockServerRunning) {
+    try {
+      await pingMockServer()
+      isMockServerRunning = true
+    } catch (err) {
+      await sleep(1000)
```

Comment:
Why do you need this sleep here?

Reply:
Aha I forgot to add some reasonable comments here. There's always an amount of delay time for the mock server to be alive completely, after the `docker run ...` command exits with the container ID. Without this waiting method, we can see several first tests to be failed, which is clearly a false positive.
Particularly the `sleep` here just to defer the `ping` action to be run every 1 second, instead of every process tick, which is unnecessarily wasteful.